java/ddlogapi.c

```diff
@@ -792,6 +792,55 @@ JNIEXPORT jlong JNICALL Java_ddlogapi_DDlogAPI_ddlog_1struct(
     return (jlong)result;
 }
 
+JNIEXPORT jlong JNICALL Java_ddlogapi_DDlogAPI_ddlog_1named_1struct(
+    JNIEnv *env, jclass obj, jstring s, jobjectArray names, jlongArray handles) {
+    const char* str = (*env)->GetStringUTFChars(env, s, NULL);
+    jsize len = (*env)->GetArrayLength(env, handles);
+    jlong *a = (*env)->GetLongArrayElements(env, handles, NULL);
+    if (a == NULL)
+        return -1;
+    jsize nameLen = (*env)->GetArrayLength(env, names);
+    if (nameLen != len) {
+        throwDDlogException(env, "number of names does not match number of values");
+        return -1;
+    }
+    ddlog_record** fields = malloc(len * sizeof(ddlog_record*));
+    if (fields == NULL) {
+        throwOutOfMemException(env, "Could not allocate buffer for %d fields.", len);
+        return -1;
+    }
+    for (size_t i = 0; i < len; i++)
+        fields[i] = (ddlog_record*)a[i];
+
+    const char** fieldNames = malloc(len * sizeof(char*));
+    if (fieldNames == NULL) {
+        free(fields);
+        throwOutOfMemException(env, "Could not allocate buffer for %d names.", len);
+        return -1;
+    }
+    for (size_t i = 0; i < len; i++) {
+        jobject jname = (*env)->GetObjectArrayElement(env, names, i);
+        if (jname == NULL) {
+            free(fields);
```

Comment:
might be cleaner to just use a goto to jump to the end of the function and call the frees you need.